File: scripts/LocationPublisher.py
#!/usr/bin/python
import rospy
from openhab_msgs.msg import LocationCommand

class LocationPublisher(object):
    """Node example class."""

    def __init__(self, item_name):

        self.item_name = item_name
        self.pub = rospy.Publisher("/openhab/items/" + self.item_name + "/command", LocationCommand, queue_size=10)
        self.rate = rospy.Rate(10) # 10hz

        # Initialize message variables.
        self.enable = False
        self.message = ""

        if self.enable:
            self.start()
        else:
            self.stop()

    def start(self):
        """Turn on publisher."""
        self.enable = True
        self.pub = rospy.Publisher("/openhab/items/" + self.item_name + "/command", LocationCommand, queue_size=10)

        while not rospy.is_shutdown():

            self.message = LocationCommand()

            self.message.latitude = 48.0501442
            self.message.longitude = 8.2014192
            self.message.altitude = 857.0

            self.message.header.stamp = rospy.Time.now()
            self.message.header.frame_id = "/base_link"
            self.message.item = self.item_name

            message = "Publishing to %s at %s: latitude = %s, longitude = %s, altitude = %s" % (self.message.item, rospy.get_time(), self.message.latitude, self.message.longitude, self.message.altitude)
            rospy.loginfo(message)

            self.pub.publish(self.message)
            self.rate.sleep()

    def stop(self):
        """Turn off publisher."""
        self.enable = False
        self.pub.unregister()

# Main function.
if __name__ == "__main__":
    # Initialize the node and name it.
    rospy.init_node("LocationPublisherNode", anonymous=True)
    # Go to class functions that do all the heavy lifting.

    locationPublisher = LocationPublisher("testLocation")

    try:
        locationPublisher.start()
    except rospy.ROSInterruptException:
        pass
    # Allow ROS to go to all callbacks.
    # spin() simply keeps python from exiting until this node is stopped
    rospy.spin()

File: src/LocationPublisher.cpp
#include "ros/ros.h"
#include "openhab_msgs/LocationCommand.h"

class LocationPublisher {
public:
    LocationPublisher(std::string item_name)
        : item_name_(item_name),
          pub_(nh_.advertise<openhab_msgs::LocationCommand>("/openhab/items/" + item_name_ + "/command", 10)),
          rate_(10)
    {
        // Initialize message variables.
        enable_ = false;
        message_ = openhab_msgs::LocationCommand();
    }

    void start() {
        // Turn on publisher.
        enable_ = true;
        pub_ = nh_.advertise<openhab_msgs::LocationCommand>("/openhab/items/" + item_name_ + "/command", 10);

        while (ros::ok()) {
            message_.latitude = 48.0501442;
            message_.longitude = 8.2014192;
            message_.altitude = 857.0;
            message_.isnull = false;
            message_.header.stamp = ros::Time::now();
            message_.header.frame_id = "/base_link";
            message_.item = item_name_;

            std::string message = "Publishing to " + message_.item + " at " + std::to_string(ros::Time::now().toSec()) + ": latitude = " + std::to_string(message_.latitude) + ", longitude = " + std::to_string(message_.longitude) + ", altitude = " + std::to_string(message_.altitude);
            ROS_INFO("%s", message.c_str());

            pub_.publish(message_);
            rate_.sleep();
        }
    }

    void stop() {
        // Turn off publisher.
        enable_ = false;
        pub_.shutdown();
    }

private:
    std::string item_name_;
    bool enable_;
    openhab_msgs::LocationCommand message_;
    ros::NodeHandle nh_;
    ros::Publisher pub_;
    ros::Rate rate_;
};

int main(int argc, char **argv) {
    // Initialize the node and name it.
    ros::init(argc, argv, "LocationPublisherNode");
    // Go to class functions that do all the heavy lifting.
    LocationPublisher location_publisher("testLocation");
    try {
        location_publisher.start();
    } catch (const ros::Exception &e) {
        ROS_ERROR("%s", e.what());
    }
    // Allow ROS to go to all callbacks.
    // spin() simply keeps C++ from exiting until this node is stopped.
    ros::spin();
    return 0;
}
